### ISSUE TITLE
DevSim Layer Needs to support extended features and properties queries

### ISSUE THREAD

**Issue by tobine** [LunarG/VulkanTools#389]
The vkGetPhysicalDeviceProperties2KHR() and vkGetPhysicalDeviceFeatures2KHR() functions add handling for extensions via the pNext tree. The DevSim layer needs to read in any extended properties/features from json file and appropriately report them when GPDP2 and GPDF2 are called.

I'm adding this support to mock driver in LVL [PR2388](https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/pull/2388) which can be used for reference.

I started to add the support to DevSim, but there's a bit of json reading that wasn't immediately obvious to me so wanted to see if @mikew-lunarg was willing to run with this.

I've pushed branch tobin_devsim_gpdp2_ext to VulkanTools repo which has one WIP [commit](https://github.com/LunarG/VulkanTools/commit/e8036f9c3447525822b3d555445a5b728e869aca) to get things started.

Not an immediate need to get the full support in, but until maxPushDescriptors is added my LVL PR is blocked. My WIP PR is basically a hard-coded hack that assigns a static value to that instead of whatever was read from json.

**Comment by mikew-lunarg** [comment]
This should be relabeled from bug to enhancement, as it really is the addition of new functionality.

**Comment by tobine** [comment]
Hi @mikew-lunarg note that existing json seems to include relevant data already. Here's one example:

https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/blob/master/tests/device_profiles/geforce_940m.json#L33

Note the block under there for the VK_KHR_push_descriptor extension:
`            {
                "extension": "VK_KHR_push_descriptor",
                "name": "maxPushDescriptors",
                "value": "32"
            },`

Five of the profiles that we currently use for testing include that data. Can we just use that as-is? I looked at that initially but wasn't sure how to parse it out with existing code as I don't do much json munging.

**Comment by mikew-lunarg** [comment]
Hi @tobine (been away, sorry for the delay)

The line you identify above is not specified by the DevSim JSON schema.   

The schema _allows_ arbitrary sections to be included in a JSON config file (in this case, the "extended" object containing the "deviceproperties2" array) but DevSim _ignores_ any such sections because their content is undefined.  A side-effect of the DevSim effort has been to formally define the data via schemas, so JSON validators can error-check the data, rather than DevSim itself.

As mentioned above, we need to create a schema for the desired extensions, extend DevSim's JSON loader to support that schema, and (in the case of gpuinfo.org) have that output validate against the new schema.

**Comment by sfricke-samsung** [comment]
@mikew-lunarg So wanting this feature for the Validation layers testing, curious where this issue stands.

> DevSim ignores any such sections because their content is undefined

Not sure what in `"extended": { "devicefeatures2": [] }` is undefined, example:

```
{
    "extension": "VK_KHR_pipeline_executable_properties",
    "name": "pipelineExecutableInfo",
    "supported": true
},
```

Also not sure (haven't looked super closely) if this is still a schema issue, devsim issue, or being able to generate from the spec issue

**Comment by mikew-lunarg** [comment]
Well it's undefined in that there's no schema defined for it, thus the
DevSim layer doesn't implement it.  Without a schema there's no way to
predict what else might appear in a given JSON object.  The schema design
leverages the naming and type conventions of Vulkan.  So while JSON can
support arbitrary content, new officially supported content needs additions
to the schema, and from the schema comes the layer implementation.

**Comment by jeremyk-lunarg** [comment]
Addressed by PR #1385 and PR #1398
